docs/keto/modeling/create-permission-model.mdx

```diff
@@ -0,0 +1,377 @@
+      //highlight-start
+      view: (ctx: Context): boolean =>
+        this.related.viewers.includes(ctx.subject) ||
+        this.related.editors.includes(ctx.subject) ||
+        this.related.owners.includes(ctx.subject) ||
+        this.related.parents.traverse((parent) => parent.permits.view(ctx)),
+      //highlight-end
+    }
+  }
+
+  class Folder implements Namespace {
+    related: {
+      owners: User[]
+      editors: User[]
+      viewers: User[]
+      parents: Folder[]
+    }
+  }
+  ```
+
+- Code for the remaining permissions:
+
+  ```ts title="permissions-v4.ts"
+  import { Namespace, Context } from "@ory/keto-namespace-types"
+
+  class User implements Namespace {}
+
+  class Document implements Namespace {
+    related: {
+      owners: User[]
+      editors: User[]
+      viewers: User[]
+      parents: Folder[]
+    }
+
+    permits = {
+      view: (ctx: Context): boolean =>
+        this.related.viewers.includes(ctx.subject) ||
+        this.related.editors.includes(ctx.subject) ||
+        this.related.owners.includes(ctx.subject) ||
+        this.related.parents.traverse((parent) => parent.permits.view(ctx)),
+      //highlight-start
+      edit: (ctx: Context): boolean =>
+        this.related.editors.includes(ctx.subject) ||
+        this.related.owners.includes(ctx.subject) ||
+        this.related.parents.traverse((parent) => parent.permits.edit(ctx)),
+      delete: (ctx: Context): boolean =>
+        this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.delete(ctx)),
+      share: (ctx: Context): boolean =>
+        this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
+      //highlight-end
+    }
+  }
+
+  class Folder implements Namespace {
+    related: {
+      owners: User[]
+      editors: User[]
+      viewers: User[]
+      parents: Folder[]
+    }
+
+    //highlight-start
+    permits = {
+      delete: (ctx: Context): boolean =>
+        this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.delete(ctx)),
+      share: (ctx: Context): boolean =>
+        this.related.owners.includes(ctx.subject) || this.related.parents.traverse((parent) => parent.permits.share(ctx)),
+    }
+    //highlight-end
+  }
+  ```
+
+#### Optional: refactoring
+
+Notice that the permission model you created is hierarchical. This means that everybody who can `view` can also `edit` documents,
+and everybody that can `edit` can also `delete` and `share`.
+
+You can refactor the `view` permission like this:
+
+```diff
+  view: (ctx: Context): boolean =>
+    this.related.viewers.includes(ctx.subject) ||
+-   this.related.editors.includes(ctx.subject) ||
+-   this.related.owners.includes(ctx.subject) ||
+-   this.related.parents.traverse((parent) => parent.permits.view(ctx))
++   this.permits.edit(ctx)
+```
+
+When you apply this to the entire config, you get this code:
```

Comment:
It's cool that you showed how to refactor one aspect of the model. How about showing how to refactor something else? Would be cool to have a more detailed look at two things before you just go BAM CONFIG. WDYT?

Reply:
There is (in the current version of the OPL at least) no more room for refactoring here.

In the future it would be nice to extract `Document` and `Folder` into a common superclass, but that needs more work in the actual Keto code.

Reply:
got it